crates/topos-tce-broadcast/src/double_echo/mod.rs

```diff
@@ -412,34 +263,165 @@ impl DoubleEcho {
+                } else {
+                    DOUBLE_ECHO_BUFFER_CAPACITY.inc();
+                    // Adding one to the pending_certificate_count because we
+                    // can't buffer it right now
+                    _ = self.pending_certificate_count.checked_add(1);
+                }
+            });
         }
     }
 
-    pub(crate) fn handle_ready(&mut self, from_peer: PeerId, certificate_id: &CertificateId) {
-        if let Some((_certificate, state)) = self.cert_candidate.get_mut(certificate_id) {
-            Self::sample_consume_peer(&from_peer, state, SampleType::ReadySubscription);
+    pub(crate) fn handle_echo(
+        &mut self,
+        from_peer: PeerId,
+        certificate_id: CertificateId,
+        ctx: Span,
+    ) {
+        let cert_delivered = self.delivered_certificates.get(&certificate_id).is_some();
+        if !cert_delivered {
+            if self.known_certificates.get(&certificate_id).is_some() {
+                let span = if let Some(root) = self.span_tracker.get(&certificate_id) {
+                    info!("DEBUG::Receive ECHO with root");
+                    info_span!(
+                        parent: root,
+                        "RECV Inbound Echo",
+                        peer = self.local_peer_id,
+                        certificate_id = certificate_id.to_string()
+                    )
+                } else {
+                    info!("DEBUG::Receive ECHO without root");
+                    info_span!(
+                        "RECV Inbound Echo",
+                        peer = self.local_peer_id,
+                        certificate_id = certificate_id.to_string()
+                    )
+                };
+
+                let _enter = span.enter();
+                debug!(
+                    "Handling DoubleEchoCommand::Echo from_peer: {} cert_id: {}",
+                    &from_peer, certificate_id
+                );
+                self.consume_echo(from_peer, &certificate_id);
+
+                self.state_change_follow_up();
+                drop(_enter);
+                // need to deliver the certificate
+            } else if self.delivered_certificates.get(&certificate_id).is_none() {
+                info!("DEBUG::Receive ECHO BUFFERING");
+                // need to buffer the Echo
+                self.buffered_messages
+                    .entry(certificate_id)
+                    .or_default()
+                    .push(DoubleEchoCommand::Echo {
+                        from_peer,
+                        certificate_id,
+                        ctx,
+                    });
+                DOUBLE_ECHO_BUFFERED_MESSAGE_COUNT.inc();
+            }
         }
     }
 
-    #[cfg_attr(feature = "direct", allow(dead_code))]
-    pub(crate) fn handle_broadcast(&mut self, cert: Certificate, origin: bool) {
-        info!("🙌 Starting broadcasting the Certificate {}", &cert.id);
+    pub(crate) fn handle_ready(
```

Comment:
Not sure about this, but since the `mod.rs` file is getting quite long, would it be helpful to move these into their own files? `handle_ready.rs`, `handle_echo.rs`. Or even `handlers.rs` with the functions in there?

Reply:
I would prefer to reserve this kind of refac for the work on the per `task` broadcast. This PR is already large enough ^^